subprojects/core/src/main/java/org/gradle/internal/fingerprint/impl/DefaultHistoricalFileCollectionFingerprint.java

```diff
@@ -25,21 +25,30 @@
 import org.gradle.api.internal.changedetection.state.mirror.logical.FingerprintCompareStrategy;
 import org.gradle.internal.fingerprint.FileCollectionFingerprint;
 import org.gradle.internal.fingerprint.HistoricalFileCollectionFingerprint;
+import org.gradle.internal.hash.HashCode;
 import org.gradle.internal.serialize.Decoder;
 import org.gradle.internal.serialize.Encoder;
+import org.gradle.internal.serialize.HashCodeSerializer;
 import org.gradle.internal.serialize.Serializer;
 
+import javax.annotation.Nullable;
 import java.io.IOException;
 import java.util.Map;
 
 public class DefaultHistoricalFileCollectionFingerprint implements HistoricalFileCollectionFingerprint {
 
     private final Map<String, NormalizedFileSnapshot> fingerprints;
     private final FingerprintCompareStrategy compareStrategy;
+    private final HashCode hashCode;
 
     public DefaultHistoricalFileCollectionFingerprint(Map<String, NormalizedFileSnapshot> fingerprints, FingerprintCompareStrategy compareStrategy) {
+        this(fingerprints, compareStrategy, null);
+    }
+
+    public DefaultHistoricalFileCollectionFingerprint(Map<String, NormalizedFileSnapshot> fingerprints, FingerprintCompareStrategy compareStrategy, @Nullable HashCode hashCode) {
         this.fingerprints = fingerprints;
         this.compareStrategy = compareStrategy;
+        this.hashCode = hashCode;
     }
```

Comment:
Again, constructors taking `null` for a `final` field is a sign that we need more fine-grained types.